File: solutions/python3/0904.py
class Solution:
    def totalFruit(self, tree: List[int]) -> int:
        ans = 0
        count = collections.Counter()

        j = 0
        for i, t in enumerate(tree):
            count[t] += 1
            while len(count) > 2:
                count[tree[j]] -= 1
                if count[tree[j]] == 0:
                    del count[tree[j]]
                j += 1
            ans = max(ans, i - j + 1)

        return ans

File: solutions/cpp/0904.cpp
class Solution {
 public:
  int totalFruit(vector<int>& tree) {
    int ans = 0;
    unordered_map<int, int> count;

    for (int i = 0, j = 0; i < tree.size(); ++i) {
      ++count[tree[i]];
      while (count.size() > 2) {
        if (--count[tree[j]] == 0) count.erase(tree[j]);
        ++j;
      }
      ans = max(ans, i - j + 1);
    }

    return ans;
  }
};